src/funnels/Depositor.sol

```diff
@@ -187,238 +105,186 @@ contract Depositor {
-        uint128 liquidity
-    ) internal pure returns (uint256 amount0, uint256 amount1) {
-        if (sqrtRatioAX96 > sqrtRatioBX96) (sqrtRatioAX96, sqrtRatioBX96) = (sqrtRatioBX96, sqrtRatioAX96);
-
-        if (sqrtRatioX96 <= sqrtRatioAX96) {
-            amount0 = _getAmount0ForLiquidityRoundingUp(sqrtRatioAX96, sqrtRatioBX96, liquidity);
-        } else if (sqrtRatioX96 < sqrtRatioBX96) {
-            amount0 = _getAmount0ForLiquidityRoundingUp(sqrtRatioX96, sqrtRatioBX96, liquidity);
-            amount1 = _getAmount1ForLiquidityRoundingUp(sqrtRatioAX96, sqrtRatioX96, liquidity);
-        } else {
-            amount1 = _getAmount1ForLiquidityRoundingUp(sqrtRatioAX96, sqrtRatioBX96, liquidity);
-        }
+    function _getLiquidityForAmts(
+        UniV3PoolLike pool,
+        int24         tickLower,
+        int24         tickUpper,
+        uint256       amt0Desired,
+        uint256       amt1Desired
+    ) internal view returns (uint128 liquidity) {
+        (uint160 sqrtPriceX96, , , , , , ) = pool.slot0();
+        uint160 sqrtRatioAX96 = TickMath.getSqrtRatioAtTick(tickLower);
+        uint160 sqrtRatioBX96 = TickMath.getSqrtRatioAtTick(tickUpper);
+
+        liquidity = LiquidityAmounts.getLiquidityForAmounts(
+            sqrtPriceX96,
+            sqrtRatioAX96,
+            sqrtRatioBX96,
+            amt0Desired,
+            amt1Desired
+        );
     }
 
-    struct DepositParams {
+    struct MintCallbackData {
         address gem0;
         address gem1;
-        uint256 amt0;
-        uint256 amt1;
-        uint256 minAmt0;
-        uint256 minAmt1;
-        uint24 fee;
-        int24 tickLower;
-        int24 tickUpper;
+        uint24  fee;
+        address payer;
     }
 
-    function _getOptimalDepositAmounts(DepositParams memory p) internal view returns (uint256 amt0, uint256 amt1) {
-        (uint160 sqrtPriceX96,,,,,,) = UniV3PoolLike(_getPoolAddress(p.gem0, p.gem1, p.fee)).slot0();
-        uint160 sqrtRatioAX96 = TickMath.getSqrtRatioAtTick(p.tickLower);
-        uint160 sqrtRatioBX96 = TickMath.getSqrtRatioAtTick(p.tickUpper);
-        uint128 liquidity = LiquidityAmounts.getLiquidityForAmounts(sqrtPriceX96, sqrtRatioAX96, sqrtRatioBX96, p.amt0, p.amt1);
-        (amt0, amt1) = _getAmountsForLiquidityRoundingUp(sqrtPriceX96, sqrtRatioAX96, sqrtRatioBX96, liquidity);
+    // https://github.com/Uniswap/v3-core/blob/d8b1c635c275d2a9450bd6a78f3fa2484fef73eb/contracts/interfaces/callback/IUniswapV3MintCallback.sol#L6
+    function uniswapV3MintCallback(
+        uint256        amt0Owed,
+        uint256        amt1Owed,
+        bytes calldata data
+    ) external {
+        MintCallbackData memory decoded = abi.decode(data, (MintCallbackData));
+        address pool = address(_getPool(decoded.gem0, decoded.gem1, decoded.fee));
+        require(msg.sender == pool, "Depositor/sender-not-a-pool");
+
+        if (amt0Owed > 0) GemLike(decoded.gem0).transferFrom(decoded.payer, msg.sender, amt0Owed);
```

Comment:
Why have `payer` as part of `MintCallbackData` instead of reading `buffer` directly here?